Allow overlapping span batch

ImTei opened the pull request:
## Context
- Current span batch spec does not allow span batches that overlaps current safe chain.
- So the op-node cannot proceed derivation if the pipeline is reset to the middle of span batch.
- It's very hard to guarantee that the op-node will always reset the safe head to the boundary of the span batch.
- Therefore we need to allow overlapping span batches.

## Changes
- Specs
  - Updated span batch validation rules to allow overlapping batches
- op-e2e test
  - Added a test case to simulate reorg scenario that should be failed on the current specs. 
    - After the L1 reorg, derivation pipeline is reset to the middle of span batch.
  - So this test case fails for now, with `OP_E2E_USE_SPAN_BATCH=true` env.

ImTei committed:
Update span batch spec to allow overlapping batch

ImTei committed:
Add e2e test case to test overlap batch after reorg

pcw109550 commented on specs/span-batches.md:
```diff
@@ -247,14 +249,31 @@ Span-batches share the same queue with v0 batches: batches are processed in L1 i
+Definitions:
+- `batch` as defined in the [Span batch format section][span-batch-format].
+- `parent_safe_block` is the L2 block from the current safe chain,
+  whose timestamp is at `span_start.timestamp - block_time`
```
```suggestion
  whose timestamp is at `span_start.timestamp - l2_block_time`
```

pcw109550 commented on specs/span-batches.md:
```diff
@@ -247,14 +249,31 @@ Span-batches share the same queue with v0 batches: batches are processed in L1 i
+
+Definitions:
+- `batch` as defined in the [Span batch format section][span-batch-format].
+- `parent_safe_block` is the L2 block from the current safe chain,
```
This block may not exist. Can you mention this? Or we may change the name of this variable, as `prev_l2_block`.

pcw109550 commented on specs/span-batches.md:
```diff
@@ -247,14 +249,31 @@ Span-batches share the same queue with v0 batches: batches are processed in L1 i
+  - Rules:
+    - For each `block_input`, whose timestamp is less than `next_timestamp`:
+      - If there's no `safe_block` for the `block_input` -> `drop`: i.e. misaligned timestamp
+      - `block_input.origin.number != safe_block.origin.number` -> `drop`
```
```suggestion
      - `block_input.l1_origin.number != safe_block.l1_origin.number` -> `drop`
```
for consistency.

pcw109550 commented on specs/span-batches.md:
```diff
@@ -247,14 +249,31 @@ Span-batches share the same queue with v0 batches: batches are processed in L1 i
+- Overlapped blocks checks:
+  - Note: If the span batch overlaps the current L2 safe chain, we must validate all overlapped blocks.
+  - Variables:
+    - `block_input`: an L2 block derived from the span-batch.
```
`block_input` redefined at `Max Sequencer time-drift checks`. We can copy and paste its definition.

pcw109550 replied:
Definition is intentionally different.

pcw109550 commented on specs/span-batches.md:
```diff
@@ -247,14 +249,31 @@ Span-batches share the same queue with v0 batches: batches are processed in L1 i
+      - If there's no `safe_block` for the `block_input` -> `drop`: i.e. misaligned timestamp
+      - `block_input.origin.number != safe_block.origin.number` -> `drop`
+      - `block_input.transactions != safe_block.transactions` -> `drop`
+        - `transactions` is a list of EIP-2718 transactions
```
No need to explain what `transactions` is. I think we can remove this line.

pcw109550 commented on specs/span-batches.md:
```diff
@@ -285,7 +301,7 @@ Span-batch rules, in validation order:
       It may reach to the next origin outside the L1 origins of the span.
   - Rules:
-    - For each `block_input` that can be read from the span-batch:
+    - For each `block_input`, whose timestamp is greater than `safe_head.timestamp`:
```
```suggestion
    - For each `block_input` whose timestamp is greater than `safe_head.timestamp`:
```

ImTei committed:
Apply code review

ImTei committed:
Update variable name

ImTei committed:
Fetch safe blocks directly in CheckBatch

ImTei committed:
Add L2BlockRefByNumber method to OracleEngine

ImTei committed:
Use new CheckBatch() in batch queue

Add unit test cases

ImTei committed:
Fix MockL2Client to return expected error

ImTei committed:
Fix batch queueing

ImTei committed:
Fix processing overlapping batch

- Compare overlapping blocks' transactions excluding deposit txs
- Fix finding L1 origin blocks from the given L1blocks in Checkbatch()
- GetSingularBatches returns excluding overlapped blocks

ImTei committed:
Add missing line

ImTei committed:
Update span-batches.md

ImTei committed:
Add old epoch check

ImTei committed:
Small fixes from code review

ImTei committed:
Fix typo

ImTei committed:
Merge pull request #24 from testinprod-io/tei/span-batch-overlap-impl

Overlapping span batch implementation
